File: backend/server/square_config.py
"""Single authoritative Square environment contract."""

from __future__ import annotations

import os
from typing import Mapping


VALID_SQUARE_ENVIRONMENTS = frozenset({"sandbox", "production"})
APPLICATION_ID_PREFIX = {
    "sandbox": "sandbox-sq0idb-",
    "production": "sq0idp-",
}
REQUIRED_CREDENTIAL_KEYS = (
    "SQUARE_ACCESS_TOKEN",
    "SQUARE_APPLICATION_ID",
    "SQUARE_LOCATION_ID",
)


class SquareConfigurationError(ValueError):
    """Raised when Square configuration is absent, ambiguous, or inconsistent."""


def resolve_square_environment(
    explicit: str | None = None,
    environ: Mapping[str, str] | None = None,
) -> str:
    """Resolve one explicit environment and reject caller/environment drift."""
    values = os.environ if environ is None else environ
    configured = values.get("SQUARE_ENVIRONMENT", "").strip().lower()
    requested = (explicit or "").strip().lower()
    if requested and configured and requested != configured:
        raise SquareConfigurationError(
            "Square environment override conflicts with SQUARE_ENVIRONMENT"
        )
    environment = requested or configured
    if environment not in VALID_SQUARE_ENVIRONMENTS:
        raise SquareConfigurationError(
            "SQUARE_ENVIRONMENT must be explicitly set to sandbox or production"
        )
    return environment


def validate_square_application_id(environment: str, application_id: str) -> None:
    """Reject frontend/backend application IDs belonging to the other environment."""
    expected_prefix = APPLICATION_ID_PREFIX[environment]
    if not application_id.startswith(expected_prefix):
        raise SquareConfigurationError(
            "SQUARE_APPLICATION_ID does not match SQUARE_ENVIRONMENT"
        )


def validate_square_configuration(
    *,
    explicit_environment: str | None = None,
    environ: Mapping[str, str] | None = None,
    require_credentials: bool = True,
) -> str:
    """Validate environment and credential completeness without exposing values."""
    values = os.environ if environ is None else environ
    environment = resolve_square_environment(explicit_environment, values)
    configured = {
        key: bool(values.get(key, "").strip()) for key in REQUIRED_CREDENTIAL_KEYS
    }
    if require_credentials and not all(configured.values()):
        missing = ", ".join(key for key, present in configured.items() if not present)
        raise SquareConfigurationError(f"Missing required Square configuration: {missing}")
    if configured["SQUARE_APPLICATION_ID"]:
        validate_square_application_id(
            environment, values["SQUARE_APPLICATION_ID"].strip()
        )
    return environment

File: frontend/src/services/squareConfig.js
export const SQUARE_SDK_URLS = Object.freeze({
  sandbox: 'https://sandbox.web.squarecdn.com/v1/square.js',
  production: 'https://web.squarecdn.com/v1/square.js'
});

const APPLICATION_ID_PREFIX = Object.freeze({
  sandbox: 'sandbox-sq0idb-',
  production: 'sq0idp-'
});

const PLACEHOLDER_APP_ID = 'sandbox-sq0idb-your-app-id';
const PLACEHOLDER_LOCATION_ID = 'your-location-id';

export function getSquareConfig(env) {
  const environment = env.VITE_SQUARE_ENVIRONMENT?.trim().toLowerCase();
  const applicationId = env.VITE_SQUARE_APPLICATION_ID?.trim();
  const locationId = env.VITE_SQUARE_LOCATION_ID?.trim();

  if (!Object.hasOwn(SQUARE_SDK_URLS, environment)) {
    throw new Error(
      'VITE_SQUARE_ENVIRONMENT must be explicitly set to sandbox or production.'
    );
  }
  if (
    !applicationId ||
    applicationId === PLACEHOLDER_APP_ID ||
    !locationId ||
    locationId === PLACEHOLDER_LOCATION_ID
  ) {
    throw new Error('Square application and location credentials are not configured.');
  }
  if (!applicationId.startsWith(APPLICATION_ID_PREFIX[environment])) {
    throw new Error(
      'VITE_SQUARE_APPLICATION_ID does not match VITE_SQUARE_ENVIRONMENT.'
    );
  }

  return {
    environment,
    applicationId,
    locationId,
    sdkUrl: SQUARE_SDK_URLS[environment],
    sandbox: environment === 'sandbox'
  };
}
